pysmt/cnf.py

```diff
@@ -0,0 +1,217 @@
+#
+# This file is part of pySMT.
+#
+#   Copyright 2014 Andrea Micheli and Marco Gario
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+import pysmt.shortcuts as shortcuts
+from pysmt.walkers import DagWalker
+from pysmt.typing import BOOL
+
+class CNFizer(DagWalker):
+
+    THEORY_PLACEHOLDER = "__Placeholder__"
+
+    TRUE_CNF = frozenset()
+    FALSE_CNF = frozenset([frozenset()])
+
+    def __init__(self, environment=None):
+        DagWalker.__init__(self)
+        self.env = environment if environment else shortcuts.get_env()
+        self.mgr = self.env.formula_manager
+        self._introduced_variables = {}
+        self._cnf_pieces = {}
+
+    def _key_var(self, formula):
+        if formula in self._introduced_variables:
+            res = self._introduced_variables[formula]
+        else:
+            res = self.mgr.FreshSymbol()
+            self._introduced_variables[formula] = res
+        return res
+
+    def convert(self, formula):
+        tl, cnf = self.walk(formula)
+        res = [frozenset([tl])]
+        for clause in cnf:
```

Comment:
This second part is not needed for correctness of convert(). Returning _res_ is enough.
Lets factor out this "simplification" part.